---
title: PA 4.1 
author: Marvin (Wenxiang) Li
format:
    html:
        toc: true
        code-fold: true
embed-resources: true        
---

# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [69]:
import pandas as pd

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [70]:
# YOUR CODE HERE
import requests
response = requests.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

Use Beautiful Soup to parse this string into a tree called `soup`

In [71]:
# YOUR CODE HERE
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [72]:
len(soup.find_all("table"))

10

We can use attributes like `class=` and `style=` to narrow down the list.

In [73]:
tables = soup.find_all("table")
for table in tables:
    print(table.attrs)

{'class': ['sidebar', 'nomobile', 'nowraplinks']}
{'class': ['wikitable']}
{'class': ['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'], 'style': 'text-align:right'}
{'class': ['wikitable']}
{'class': ['wikitable'], 'style': 'text-align:left'}
{'class': ['sortable', 'wikitable', 'static-row-numbers', 'sort-under', 'col1left'], 'style': 'text-align:right'}
{'class': ['wikitable', 'sortable', 'sort-under']}
{'class': ['wikitable', 'sortable', 'sort-under', 'col1left', 'col2center'], 'style': 'text-align:right'}
{'class': ['nowraplinks', 'mw-collapsible', 'mw-collapsed', 'navbox-inner'], 'style': 'border-spacing:0;background:transparent;color:inherit'}
{'class': ['navbox-columns-table'], 'style': 'border-spacing: 0px; text-align:left;width:auto; margin-left:auto; margin-right:auto;'}


In [74]:
len(soup.find_all("table", attrs={"class": ['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'],
                                  'style': 'text-align:right'}))

3

In [ ]:
tables = soup.find_all("table", attrs={"class": ['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'],
                                  'style': 'text-align:right'})
for table in tables:
    print(table.attrs)

{'class': ['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'], 'style': 'text-align:right'}
{'class': ['sortable', 'wikitable', 'static-row-numbers', 'sort-under', 'col1left'], 'style': 'text-align:right'}
{'class': ['wikitable', 'sortable', 'sort-under', 'col1left', 'col2center'], 'style': 'text-align:right'}


At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [75]:
table = soup.find_all("table", attrs={"class": ['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'],
                                  'style': 'text-align:right'})[0]

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

- test out code

In [76]:
text_row = table.find_all("tr")[2]
cell_test = text_row.find_all("td")
cell_test[0]

<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
</td>

In [77]:
city = cell_test[0].find("a")
city_name = city.text.strip()
city_name

'New York'

In [78]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED
rows = []

# iterate over all rows in the faculty table
for city in table.find_all("tr")[2:]:

  # Get all the cells (<td>) in the row.
    cells = city.find_all("td")

    # Find the the name of the city in cell[0]
    # city is contained in the <a> tag
    city = cells[0].find("a") or cells[0]
    city_name = city.text.strip()

    # Find the state of the city in cell[1]
    # state is contained in the <a> tag
    state = cells[1].find("a") or cells[1]
    state_short = state.text.strip()

    # Find the 2022 estimate in cell[2]
    # 2022 estimate is contained in the <td> tag
    estimate = cells[2].find("td") or cells[2]
    est_22 = estimate.text.strip()

    # Find the 2022 estimate in cell[2]
    # 2022 estimate is contained in the <td> tag
    land = cells[5].find("td") or cells[5]
    land_area = land.text.strip()

    # Append this data.
    rows.append({
        "city": city_name,
        "state": state_short,
        "2022_estimate": est_22,
        "land_area": land_area
    })



In [79]:
type(rows)

list

In [141]:
wiki = pd.DataFrame(rows)
wiki.head()

city state 2022_estimate land_area
0     New York    NY     8,258,035     300.5
1  Los Angeles    CA     3,820,914     469.5
2      Chicago    IL     2,664,452     227.7
3      Houston    TX     2,314,157     640.4
4      Phoenix    AZ     1,650,070     518.0

In [142]:
wiki.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336 entries, 0 to 335
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   city           336 non-null    object
 1   state          336 non-null    object
 2   2022_estimate  336 non-null    object
 3   land_area      336 non-null    object
dtypes: object(4)
memory usage: 10.6+ KB


Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [ ]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [82]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")[2]

City  ST 2023 estimate 2020 census  Change 2020 land area          \
            City  ST 2023 estimate 2020 census  Change            mi2     km2   
0    New York[c]  NY       8258035     8804190  −6.20%          300.5   778.3   
1    Los Angeles  CA       3820914     3898747  −2.00%          469.5  1216.0   
2        Chicago  IL       2664452     2746388  −2.98%          227.7   589.7   
3        Houston  TX       2314157     2304580  +0.42%          640.4  1658.6   
4        Phoenix  AZ       1650070     1608139  +2.61%          518.0  1341.6   
..           ...  ..           ...         ...     ...            ...     ...   
331         Yuma  AZ        100858       95548  +5.56%          120.7   312.6   
332  New Bedford  MA        100695      101079  −0.38%           20.0    51.8   
333   Suffolk[l]  VA        100659       94324  +6.72%          399.2  1033.9   
334     Hesperia  CA        100633       99818  +0.82%           72.7   188.3   
335    Davenport  IA        100354      101724  −1.35%           63.8   165.2   

    2020 density                                      Location  
           / mi2  / km2                               Location  
0          29298  11312    40°40′N 73°56′W / 40.66°N 73.94°W  
1           8304   3206  34°01′N 118°25′W / 34.02°N 118.41°W  
2          12061   4657    41°50′N 87°41′W / 41.84°N 87.68°W  
3           3599   1390    29°47′N 95°23′W / 29.79°N 95.39°W  
4           3105   1199  33°34′N 112°05′W / 33.57°N 112.09°W  
..           ...    ...                                    ...  
331          792    306  32°31′N 114°31′W / 32.52°N 114.52°W  
332         5054   1951    41°40′N 70°56′W / 41.66°N 70.94°W  
333          236     91    36°42′N 76°38′W / 36.70°N 76.63°W  
334         1373    530  34°24′N 117°19′W / 34.40°N 117.32°W  
335         1594    615    41°34′N 90°36′W / 41.56°N 90.60°W  

[336 rows x 10 columns]

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [83]:
df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")[2]
df_cities2

City  ST 2023 estimate 2020 census  Change 2020 land area          \
            City  ST 2023 estimate 2020 census  Change            mi2     km2   
0    New York[c]  NY       8258035     8804190  −6.20%          300.5   778.3   
1    Los Angeles  CA       3820914     3898747  −2.00%          469.5  1216.0   
2        Chicago  IL       2664452     2746388  −2.98%          227.7   589.7   
3        Houston  TX       2314157     2304580  +0.42%          640.4  1658.6   
4        Phoenix  AZ       1650070     1608139  +2.61%          518.0  1341.6   
..           ...  ..           ...         ...     ...            ...     ...   
331         Yuma  AZ        100858       95548  +5.56%          120.7   312.6   
332  New Bedford  MA        100695      101079  −0.38%           20.0    51.8   
333   Suffolk[l]  VA        100659       94324  +6.72%          399.2  1033.9   
334     Hesperia  CA        100633       99818  +0.82%           72.7   188.3   
335    Davenport  IA        100354      101724  −1.35%           63.8   165.2   

    2020 density                                      Location  
           / mi2  / km2                               Location  
0          29298  11312    40°40′N 73°56′W / 40.66°N 73.94°W  
1           8304   3206  34°01′N 118°25′W / 34.02°N 118.41°W  
2          12061   4657    41°50′N 87°41′W / 41.84°N 87.68°W  
3           3599   1390    29°47′N 95°23′W / 29.79°N 95.39°W  
4           3105   1199  33°34′N 112°05′W / 33.57°N 112.09°W  
..           ...    ...                                    ...  
331          792    306  32°31′N 114°31′W / 32.52°N 114.52°W  
332         5054   1951    41°40′N 70°56′W / 41.66°N 70.94°W  
333          236     91    36°42′N 76°38′W / 36.70°N 76.63°W  
334         1373    530  34°24′N 117°19′W / 34.40°N 117.32°W  
335         1594    615    41°34′N 90°36′W / 41.56°N 90.60°W  

[336 rows x 10 columns]

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [ ]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

[                                        Program name   Program type
 0                              Actuarial Preparation          Minor
 1  Cross Disciplinary Studies Minor in Bioinforma...          Minor
 2   Cross Disciplinary Studies Minor in Data Science          Minor
 3                                         Statistics  BS, MS, Minor]


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [84]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED
response_2 = requests.get("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

In [85]:
soup_2 = BeautifulSoup(response_2.content, "html.parser")

In [138]:
row_2 = []
for course in soup_2.find_all("div",{"class": "courseblock"})[0:]:

    course_tag = course.find('p', class_='courseblocktitle')
    course_name = course_tag.get_text(strip = True).replace("\xa0", "")

    term_tag = course.find_all('p', class_='noindent')[0]
    term_name = term_tag.text

    row_2.append({
        'course_name': course_name,
        'term' : term_name
    })

In [140]:
calpoly = pd.DataFrame(row_2)
calpoly.head()

course_name  \
0            DATA100. Data Science for All I.4 units   
1      DATA301. Introduction to Data Science.4 units   
2   DATA401. Data Science Process and Ethics.3 units   
3  DATA402. Mathematical Foundations of Data Scie...   
4   DATA403. Data Science Projects Laboratory.1 unit   

                               term  
0  Term Typically Offered: F, W, SP  
1  Term Typically Offered: F, W, SP  
2         Term Typically Offered: F  
3         Term Typically Offered: F  
4         Term Typically Offered: F

In [143]:
calpoly.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74 entries, 0 to 73
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   course_name  74 non-null     object
 1   term         74 non-null     object
dtypes: object(2)
memory usage: 1.3+ KB


Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag